### DISCUSSION TITLE
Dynamic route rendering content dynamically

### DISCUSSION THREAD

**Discussion by john45321** [vercel/next.js#17346] [Help] [answered]
Hi, members

I have been stuck with this problem for a long time. I have 3 pages.

- pages/index.js
- pages/categories.js
- pages/categories/[slug].js

The `categories/[slug].js` is using [Next.js](https://nextjs.org/) fetching method name [getServerSideProps](https://nextjs.org/docs/basic-features/data-fetching#getserversideprops-server-side-rendering) that runs on each request and used for build dynamic pages on runtime. The `categories/[slug].js` is rendering a dynamic content on the page, but the dynamic content comes from the CMS as a response from the API Endpoint. Dynamic content is nothing but a string that contains HTML.

![response-come-from-CMS](https://user-images.githubusercontent.com/59892334/94192568-92ead800-fec8-11ea-86ed-cc63193d12c6.jpg)

**Note:** To fetch the content from the CMS we have to send a POST request with the CMS credentials like username, password, and the page slug for the content. I am using [axios ](https://github.com/axios/axios) library to send a post request and the method is inside `post.js` file.

**post.js:**

```js
import axios from 'axios';

const postCMS = async (slug) => {
    const url = `${process.env.CMS_API_URL}/render-page/`;
    let pageSlug = slug;

    // If the pageSlug is not start with `/`, then create the slug with `/`
    if (!pageSlug.startsWith('/')) {
        pageSlug = `/${pageSlug}`;
    }

    const head = {
        Accept: 'application/json',
        'Content-Type': 'application/json'
    };

    const data = JSON.stringify({
        username: process.env.CMS_API_USERNAME,
        password: process.env.CMS_API_PASSWORD,
        slug: pageSlug
    });

    try {
        const response = await axios.post(url, data, {
            headers: head
        });
        return response.data;
    } catch (e) {
        return e;
    }
};

export default postCMS;
```
But for the rendering content on the `categories/[slug].js` page, I am using the [Reactjs ](https://reactjs.org/) prop name [dangerouslySetInnerHTML](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml) to render all the HTML on the page with something like this.

**pages/categories/[slug].js**:

`<div dangerouslySetInnerHTML={{ __html: result.html }} /> `

The content is loading fine based on each slug. But when I click the category link which is located in the `categories/index.js` 

```js
<Link href="/categories/[slug]" as="/categories/online-cloud-storage">
      <a>Online Cloud Storage</a>
</Link>
```

The dynamic content is load successfully but that dynamic content contains `accordion` and `slider` elements they are not working. But when I refresh the page they work fine. See this.

![preview-problem](https://user-images.githubusercontent.com/59892334/94193754-2cff5000-feca-11ea-967f-b146aa7ae964.gif)

They also work fine when I set the Link something like this.

```
<Link href="/categories/online-cloud-storage" as="/categories/online-cloud-storage">
     <a>Online Cloud Storage</a>
 </Link>
```

But after setting the link like the above method, the click is caused to hard reload the page. But I don't want this. Everything should work. When the user clicks on the category link.

 **Is there a way to fix this**?
 **Why the content elements are not working when you click from the `categories/index.js` page**?

[Github repo](https://github.com/john45321/next.js-dynamic-content)

**Code:**

**pages/index.js**:

```js
import React from 'react';
import Link from 'next/link';

const IndexPage = () => {
    return (
        <div>
            <Link href="/categories">
                <a>Categories</a>
            </Link>
        </div>
    );
};

export default IndexPage;
```

**pages/categories/index.js**:

```js
import React from 'react';
import Link from 'next/link';

const Categories = () => {
    return (
        <div>
            <Link href="/categories/[slug]" as="/categories/online-cloud-storage">
                <a>Online Cloud Storage</a>
            </Link>
        </div>
    );
};

export default Categories;
```

**pages/categories/[slug].js**:

```js
import React from 'react';
import Head from 'next/head';
import postCMS from '../../post';

const CategoryPage = ({ result }) => {
    return (
        <>
            <Head>
                {result && <link href={result.assets.stylesheets} rel="stylesheet" />}
            </Head>
            <div>
                <h1>Category page CMS Content</h1>
                <div dangerouslySetInnerHTML={{ __html: result.html }} />
            </div>
        </>
    );
};


export const getServerSideProps = async (context) => {
  const categorySlug = context.query.slug;
  const result = await postCMS(categorySlug);
  return {
    props: {
      result
    }
  };
};


export default CategoryPage;
```

**Comment by jamesmosier** [comment, marked as answer]
I assume the accordion and other components are using javascript that is injected from the CMS? If so, [`dangerouslySetInnerHTML`](https://github.com/facebook/react/issues/8838) does not work with `<script>` tags.

**Reply by john45321** [reply]
Thanks, @jamesmosier for your great time. Yes, it contains scripts. But When I refresh the page it works fine.

**Reply by jamesmosier** [reply]
It works when reloading because that is server rendered, therefore the script tag exists when the DOM is rendered. When you navigate to it via `<Link>` it is client rendered and `dangerouslySetInnerHTML` is at its core just doing an `element.innerHTML = 'your content'` which [does not work without eval](https://stackoverflow.com/a/1197585/1554860)

**Reply by john45321** [reply]
Your answers are really grateful and improving my knowledge. So pass the whole `result.html` object in the eval function. Can you please guide me? What changes I need to do in the `pages/categories/[slug].js` page.

**Reply by jamesmosier** [reply]
You can't pass the whole result to `eval` because it is HTML and JavaScript in `result.html`. Also `eval` is super dangerous to use as it lets any script run on your page with complete control over the page.

The only way I can think you could do this would be to parse the `result.html` with regex, search for `<script>` tags, extract the contents, insert `eval` into the contents, then put it all back together and render that back to the html.

I won't be able to help to that extent unfortunately.

**Reply by john45321** [reply]
Is getInitialProps can leverage this?

**Reply by jamesmosier** [reply]
No, that won't help. You could just change your `<Link>` to an `<a>` so then it always does a full reload when you hit the page.

**Reply by john45321** [reply]
Thanks, @jamesmosier for your time.

**Reply by just-abdullah-dev** [reply]
@jamesmosier as <a> tag solve the problem when user navigate to the page. What if user goes back to the previous page. (In my case script on previous page not working but works on reload.)
